In [1]:
%load_ext bigquery_magics

# "Will it snow tomorrow?" - The time traveler asked
The following dataset contains climate information from over 9000 stations accross the world. The overall goal of these subtasks will be to predict whether it will snow tomorrow 20 years ago. So if today is 1 April 2025 then the weather we want to forecast is for the 2 April 2005. You are supposed to solve the tasks using Big Query, which can be used in the Jupyter Notebook like it is shown in the following cell. For further information and how to use BigQuery in Jupyter Notebook refer to the Google Docs. 

The goal of this test is to test your coding knowledge in Python, BigQuery and Pandas as well as your understanding of Data Science. If you get stuck in the first part, you can use the replacement data provided in the second part.

In [2]:
%%bigquery --project coding-challenge-459917
SELECT
*,
FROM `bigquery-public-data.samples.gsod`
LIMIT 20 


Query is running:   0%|          |

Downloading:   0%|          |

station_number  wban_number  year  month  day  mean_temp  \
0            36010        99999  1929     12    6  49.500000   
1            36010        99999  1929     10    7  53.200001   
2            38940        99999  1929     10   14  53.000000   
3            33790        99999  1929     10   18  46.799999   
4            30750        99999  1929     11   24  46.700001   
5            30050        99999  1929     12    1  45.700001   
6            38940        99999  1929     11   29  54.200001   
7            33790        99999  1929     10    3  52.500000   
8            33110        99999  1929     10   28  52.200001   
9            32620        99999  1929     11    3  44.599998   
10           39800        99999  1929     11   24  46.099998   
11           34970        99999  1929     11   15  30.600000   
12           38040        99999  1929     10   11  57.000000   
13           34970        99999  1929     10   10  49.400002   
14           38560        99999  1930     12   11  49.700001   
15           30050        99999  1930      2   20  41.500000   
16           37950        99999  1930     12   31  45.200001   
17           37770        99999  1930     10   24  46.299999   
18           37950        99999  1930     12    3  40.799999   
19           33790        99999  1930      2    2  41.000000   

    num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
0                       4       47.500000                           4   
1                       4       49.099998                           4   
2                       4       50.700001                           4   
3                       4       41.500000                           4   
4                       4       43.200001                           4   
5                       4       42.799999                           4   
6                       4       51.799999                           4   
7                       4       39.299999                           4   
8                       4       47.299999                           4   
9                       5       40.599998                           4   
10                      5       39.700001                           4   
11                      5       28.299999                           4   
12                      5       54.200001                           4   
13                      5       44.200001                           4   
14                      4       45.299999                           4   
15                      4       36.000000                           4   
16                      4       43.500000                           4   
17                      4       39.200001                           4   
18                      4       38.200001                           4   
19                      4       40.000000                           4   

    mean_sealevel_pressure  ...  min_temperature  min_temperature_explicit  \
0               983.200012  ...              NaN                      <NA>   
1              1002.299988  ...              NaN                      <NA>   
2              1030.699951  ...              NaN                      <NA>   
3              1009.200012  ...              NaN                      <NA>   
4               989.000000  ...              NaN                      <NA>   
5               991.599976  ...              NaN                      <NA>   
6              1001.000000  ...              NaN                      <NA>   
7               998.599976  ...              NaN                      <NA>   
8              1003.400024  ...              NaN                      <NA>   
9              1021.200012  ...              NaN                      <NA>   
10              985.799988  ...              NaN                      <NA>   
11             1002.299988  ...              NaN                      <NA>   
12             1029.400024  ...              NaN                      <NA>   
13             1021.799988  ...  

## Part 1

### 1. Task
Change the date format to 'YYYY-MM-DD' and select the data from 2000 till 2005 for station numbers including and between 725300 and 726300 , and save it as a pandas dataframe. Note the maximum year available is 2010. 

Here I am further filtering the table to select the records where
- year is between 2000 and 2005, and
- station numbers is between 725300 and 725330.

In [3]:
%%bigquery --project coding-challenge-459917
SELECT 
*,
FROM `bigquery-public-data.samples.gsod`
WHERE (year BETWEEN 2000 AND 2005)
  AND (station_number BETWEEN 725300 AND 726300)
LIMIT 20; 

Query is running:   0%|          |

Downloading:   0%|          |

station_number  wban_number  year  month  day  mean_temp  \
0           725868        99999  2000      4    7  37.900002   
1           725835        99999  2000      1   10  47.299999   
2           725786        99999  2000      6   11  69.800003   
3           725827        99999  2000      9    6  44.200001   
4           725827        99999  2000     11    7  17.100000   
5           725786        99999  2000      2   29  41.700001   
6           725985        99999  2000      9   19  82.400002   
7           725468        99999  2000      1   31  14.900000   
8           726064        99999  2000     11    2  47.299999   
9           725826        99999  2000      3   17  35.299999   
10          725869        99999  2000      9    4  54.799999   
11          725985        99999  2000     11   12  44.000000   
12          725835        99999  2000     11    3  47.400002   
13          725835        99999  2000     11    9  34.099998   
14          725786        99999  2000     11   27  37.400002   
15          725826        99999  2000      2   13  34.599998   
16          725868        99999  2000      3    1  33.000000   
17          725985        99999  2000      2   25  49.000000   
18          725827        99999  2000      2   13  28.700001   
19          725985        99999  2000     12   25  49.000000   

    num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
0                       4       21.200001                           4   
1                       4       26.600000                           4   
2                       4       45.000000                           4   
3                       4       25.200001                           4   
4                       4       -7.200000                           4   
5                       5       33.400002                           5   
6                       5       55.799999                           5   
7                       6        9.200000                           6   
8                       6       29.600000                           6   
9                       6       30.200001                           6   
10                      6       38.900002                           6   
11                      6       35.900002                           6   
12                      7       23.799999                           7   
13                      7       26.299999                           7   
14                      7       31.700001                           7   
15                      7       32.799999                           7   
16                      7       28.900000                           7   
17                      7       42.500000                           7   
18                      7       11.700000                           7   
19                      7       37.400002                           7   

    mean_sealevel_pressure  ...  min_temperature  min_temperature_explicit  \
0                      NaN  ...              NaN                      <NA>   
1                      NaN  ...              NaN                      <NA>   
2                      NaN  ...              NaN                      <NA>   
3              1018.400024  ...              NaN                      <NA>   
4              1019.400024  ...              NaN                      <NA>   
5                      NaN  ...              NaN                      <NA>   
6              1014.000000  ...              NaN                      <NA>   
7                      NaN  ...              NaN                      <NA>   
8                      NaN  ...              NaN                      <NA>   
9                      NaN  ...              NaN                      <NA>   
10                     NaN  ...              NaN                      <NA>   
11             1025.599976  ...              NaN                      <NA>   
12                     NaN  ...              NaN                      <NA>   
13                     NaN  ...  

### 2. Task 
From here you want to work with the data from all stations 725300 to 725330 that have information from 2000 till 2005. 

Here I am further filtering the table to select the records where
- year is between 2000 and 2005, and
- station numbers is between 725300 and 725330.

In [5]:
%%bigquery --project coding-challenge-459917
SELECT 
*,
FROM `bigquery-public-data.samples.gsod`
WHERE (year BETWEEN 2000 AND 2005)
  AND (station_number BETWEEN 725300 AND 725330)
LIMIT 20;

Query is running:   0%|          |

Downloading:   0%|          |

station_number  wban_number  year  month  day  mean_temp  \
0           725327        99999  2000      3   17  31.000000   
1           725326        99999  2000     12    6  13.700000   
2           725316        99999  2000      5   12  76.800003   
3           725317        99999  2000      8   26  70.300003   
4           725305        99999  2000      9   15  55.400002   
5           725317        99999  2000      4   13  47.200001   
6           725316        99999  2000      1   26  12.100000   
7           725327        99999  2000     12    5  24.000000   
8           725327        99999  2000     11    1  59.599998   
9           725314        99999  2000      6   25  75.599998   
10          725330        14827  2000     10   26  63.900002   
11          725315        99999  2000      7    6  77.599998   
12          725317        99999  2000      4    2  55.799999   
13          725320        14842  2000      6   13  74.199997   
14          725317        99999  2000      6   14  76.500000   
15          725317        99999  2000      3    5  51.700001   
16          725305        99999  2000      7   13  73.300003   
17          725305        99999  2000      6    1  69.500000   
18          725330        14827  2000     11   17  30.900000   
19          725305        99999  2000     12   19  14.100000   

    num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
0                      21       20.900000                          21   
1                      22        2.400000                          22   
2                      23       69.400002                          23   
3                      23       66.400002                          23   
4                      23       41.799999                          23   
5                      23       29.299999                          23   
6                      24        4.400000                          24   
7                      24       11.700000                          24   
8                      24       39.500000                          24   
9                      24       67.699997                          24   
10                     24       59.900002                          24   
11                     24       72.300003                          24   
12                     24       41.099998                          24   
13                     24       69.900002                          24   
14                     24       68.800003                          24   
15                     24       28.600000                          24   
16                     24       63.200001                          24   
17                     24       64.800003                          24   
18                     24       25.799999                          24   
19                     24        9.100000                          24   

    mean_sealevel_pressure  ...  min_temperature  min_temperature_explicit  \
0              1028.300049  ...              NaN                      <NA>   
1                      NaN  ...              NaN                      <NA>   
2                      NaN  ...              NaN                      <NA>   
3              1013.200012  ...              NaN                      <NA>   
4              1019.700012  ...              NaN                      <NA>   
5              1026.099976  ...              NaN                      <NA>   
6                      NaN  ...              NaN                      <NA>   
7              1023.500000  ...              NaN                      <NA>   
8              1020.000000  ...              NaN                      <NA>   
9              1014.200012  ...              NaN                      <NA>   
10             1020.700012  ...              NaN                      <NA>   
11             1013.000000  ...              NaN                      <NA>   
12             1011.700012  ...              NaN                      <NA>   
13             1011.500000  ...  

Start by checking which year received the most snowfall in our data. 

To find the year that received the most snowfall across all stations:
- first group the data by year using the SQL clause `GROUP BY`, then
- count for number of distinct days it snowed using the clauses `COUNT` and `DISTINCT`

In [7]:
%%bigquery --project coding-challenge-459917
SELECT year, COUNT(DISTINCT DATE(year, month, day)) AS total_snowed
FROM `bigquery-public-data.samples.gsod`
WHERE snow = TRUE
GROUP BY year
ORDER BY total_snowed DESC;

Query is running:   0%|          |

Downloading:   0%|          |

year  total_snowed
0   1936           366
1   1940           366
2   1944           366
3   1948           366
4   1952           366
..   ...           ...
77  1932           272
78  1931           254
79  1930           194
80  2010           105
81  1929            61

[82 rows x 2 columns]

According to the above solution, the year 1936 received the most snowfall. (However, I am somewhat skeptical about the total_snowed count. If I time permits, I plan to revisit this later.)

Add an additional field that indicates the daily change in snow depth measured at every station. And identify the station and day for which the snow depth increased the most.  

To calculate the daily change in snow depth at each station, we need to subtract the snow depth measured at each station on the previous day. This is achieved by using the function `LAG(...)` with `OVER (PARTITION BY ...)` to group by the station numbers and `ORDER BY` to currectly order the dates.

In [8]:
%%bigquery --project coding-challenge-459917
SELECT station_number, snow_depth,
 DATE(year, month, day) AS y_m_d,
 snow_depth - LAG(snow_depth) OVER (PARTITION BY station_number ORDER BY DATE(year, month, day)) AS diff
FROM `bigquery-public-data.samples.gsod`
ORDER BY diff DESC;


Query is running:   0%|          |

Downloading:   0%|          |

station_number  snow_depth       y_m_d        diff
0                  315320  117.300003  1995-04-21  116.100003
1                  153450  117.300003  1992-02-28  115.700003
2                  151600  117.300003  1991-02-20  114.900003
3                  345100  110.199997  1989-02-03  108.199997
4                  296120  109.400002  1993-10-31  107.400002
...                   ...         ...         ...         ...
114420311          110510         NaN  2002-11-28         NaN
114420312          110510         NaN  2005-04-13         NaN
114420313          110510         NaN  2008-06-04         NaN
114420314          110510         NaN  2010-03-14         NaN
114420315          112950         NaN  2002-03-09         NaN

[114420316 rows x 4 columns]

The highest increase in snow occured at the station 315320 on 1995-04-21.

Do further checks on the remaining dataset, clean or drop data depending on how you see appropriate. 

I will postpone the further checks and Task 3 (data splitting) to Part 2 using the provided csv file.

### 3. Task
Now it is time to split the data, into a training, evaluation and test set. As a reminder, the date we are trying to predict snow fall for should constitute your test set.

In [ ]:
import datetime as dt

str(dt.datetime.today()- dt.timedelta(days=20*365)).split(' ')[0]


## Part 2
If you made it up to here all by yourself, you can use your prepared dataset to train an algorithm of your choice to forecast whether it will snow on the following date for each station in this dataset:

In [ ]:
import datetime as dt

str(dt.datetime.today()- dt.timedelta(days=20*365)).split(' ')[0]

You are allowed to use any library you are comfortable with such as sklearn, tensorflow, keras etc. 
If you did not manage to finish part one feel free to use the data provided in 'coding_challenge.csv' Note that this data does not represent a solution to Part 1. 